src/main/php/com/amazon/aws/ServiceEndpoint.class.php

```diff
@@ -128,28 +128,45 @@ public function connecting($connections) {
   }
 
   /**
-   * Returns a new resource consisting of path including
-   * optional placeholders and replacement segments.
+   * Returns a new resource consisting of path including optional placeholders
+   * and replacement segments.
    * 
+   * @param  string|com.amazon.aws.S3Key $path
    * @throws lang.ElementNotFoundException
    */
-  public function resource(string $path, array $segments= []): Resource {
+  public function resource($path, array $segments= []): Resource {
     return new Resource($this, $path, $segments, $this->marshalling);
   }
 
+  /**
+   * Extracts path, encoded and params from a given target. Handles S3 keys, which do
+   * not double-encode the path component in the canonical request.
+   *
+   * @param  com.amazon.aws.api.SignatureV4 $signature
+   * @param  string|com.amazon.aws.S3Key $target
+   * @return var[]
+   */
+  private function target($signature, $target) {
+    if ($target instanceof S3Key) {
+      $path= $target->path($this->base);
+      return [$path, $signature->encoded($path), []];
+    } else if (false === ($p= strpos($target, '?'))) {
+      $path= $path= $this->base.ltrim($target, '/');
+      return [$path, $path, []];
+    } else {
+      parse_str(substr($target, $p + 1), $params);
+      $path= $this->base.ltrim(substr($target, 0, $p), '/');
+      return [$path, $path, $params];
+    }
+  }
+
   /** Signs a given target (optionally including parameters) with a given expiry time */
-  public function sign(string $target, int $expires= 3600, $time= null): string {
+  public function sign($target, int $expires= 3600, $time= null): string {
```

Comment:
With PHP 8.0+, we could use `string|S3Key` here, see https://wiki.php.net/rfc/union_types_v2